File: imports/generate_templates.py
#!/usr/bin/env python3
"""
Generate Excel import template files for all entity types.
Run once to create templates in imports/templates/.
"""

from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
import os

TEMPLATES_DIR = os.path.join(os.path.dirname(__file__), "templates")
os.makedirs(TEMPLATES_DIR, exist_ok=True)

# Styles
HEADER_FONT = Font(bold=True, size=11)
HEADER_FILL = PatternFill(fill_type="solid", fgColor="4472C4")
HEADER_FONT_WHITE = Font(bold=True, size=11, color="FFFFFF")
EXAMPLE_FILL = PatternFill(fill_type="solid", fgColor="D6E4F0")
EXAMPLE_FONT = Font(italic=True, color="333333")
DESC_FILL = PatternFill(fill_type="solid", fgColor="E2EFDA")
DESC_FONT = Font(size=10, color="555555")
ENUM_FILL = PatternFill(fill_type="solid", fgColor="D9D9D9")
ENUM_FONT = Font(size=10, color="666666", italic=True)
THIN_BORDER = Border(
    left=Side(style="thin", color="B4B4B4"),
    right=Side(style="thin", color="B4B4B4"),
    top=Side(style="thin", color="B4B4B4"),
    bottom=Side(style="thin", color="B4B4B4"),
)


def create_template(filename, columns):
    """
    Create an Excel template with 4 header rows.
    columns: list of dicts with keys: name, example, description, allowed_values
    """
    wb = Workbook()
    ws = wb.active
    ws.title = "Import"

    for col_idx, col in enumerate(columns, start=1):
        # Row 1: Header
        cell = ws.cell(row=1, column=col_idx, value=col["name"])
        cell.font = HEADER_FONT_WHITE
        cell.fill = HEADER_FILL
        cell.alignment = Alignment(horizontal="center", wrap_text=True)
        cell.border = THIN_BORDER

        # Row 2: Example data
        cell = ws.cell(row=2, column=col_idx, value=col["example"])
        cell.font = EXAMPLE_FONT
        cell.fill = EXAMPLE_FILL
        cell.border = THIN_BORDER

        # Row 3: Field descriptions
        cell = ws.cell(row=3, column=col_idx, value=col["description"])
        cell.font = DESC_FONT
        cell.fill = DESC_FILL
        cell.alignment = Alignment(wrap_text=True)
        cell.border = THIN_BORDER

        # Row 4: Allowed values (light grey)
        allowed = col.get("allowed_values", "")
        cell = ws.cell(row=4, column=col_idx, value=allowed)
        cell.font = ENUM_FONT
        cell.fill = ENUM_FILL
        cell.alignment = Alignment(wrap_text=True)
        cell.border = THIN_BORDER

        # Set column width based on content
        max_len = max(
            len(str(col["name"])),
            len(str(col["example"])),
            min(len(str(col["description"])), 40),
            min(len(str(allowed)), 40) if allowed else 0,
        )
        ws.column_dimensions[ws.cell(row=1, column=col_idx).column_letter].width = max(max_len + 4, 15)

    # Freeze top 4 rows so data entry starts below
    ws.freeze_panes = "A5"

    filepath = os.path.join(TEMPLATES_DIR, filename)
    wb.save(filepath)
    print(f"  Created: {filepath}")


# ============================================================
# Template definitions — columns match database schema exactly
# ============================================================

ENTITIES = [
    {
        "name": "entity_type",
        "example": "company",
        "description": "Required. Type of entity.",
        "allowed_values": "company | individual",
    },
    {
        "name": "document_type",
        "example": "RUC",
        "description": "Required. Type of identification document.",
        "allowed_values": "RUC | DNI | CE | Pasaporte",
    },
    {
        "name": "document_number",
        "example": "20612345678",
        "description": "Required. The actual ID number. RUC=11 digits, DNI=8 digits. Must be unique.",
        "allowed_values": "",
    },
    {
        "name": "legal_name",
        "example": "Constructora Los Andes S.A.C.",
        "description": "Required. Razón social or full legal name.",
        "allowed_values": "",
    },
    {
        "name": "common_name",
        "example": "Los Andes",
        "description": "Optional. How you refer to them day to day.",
        "allowed_values": "",
    },
    {
        "name": "city",
        "example": "Arequipa",
        "description": "Optional. City where the entity is located. Enables geographic filtering.",
        "allowed_values": "",
    },
    {
        "name": "region",
        "example": "Arequipa",
        "description": "Optional. Peruvian department/region.",
        "allowed_values": "",
    },
    {
        "name": "notes",
        "example": "Cement supplier from Arequipa",
        "description": "Optional. Free text notes.",
        "allowed_values": "",
    },
]

PROJECTS = [
    {
        "name": "project_code",
        "example": "PRY001",
        "description": "Optional. If blank, auto-generated as next sequential PRY###. If provided, must be unique and follow PRY### format.",
        "allowed_values": "PRY### format",
    },
    {
        "name": "name",
        "example": "Pavimentación Av. Los Héroes",
        "description": "Required. Project name.",
        "allowed_values": "",
    },
    {
        "name": "project_type",
        "example": "subcontractor",
        "description": "Required. Type of project.",
        "allowed_values": "subcontractor | oxi",
    },
    {
        "name": "status",
        "example": "active",
        "description": "Required. Current project status.",
        "allowed_values": "prospect | active | completed | cancelled",
    },
    {
        "name": "client_entity_document_number",
        "example": "20612345678",
        "description": "Optional. Document number of the client entity. Must exist in entities table. Nullable for prospects.",
        "allowed_values": "Lookup → entities.document_number",
    },
    {
        "name": "contract_value",
        "example": "500000.00",
        "description": "Optional. Total contract value. NUMERIC(15,2).",
        "allowed_values": "",
    },
    {
        "name": "contract_currency",
        "example": "PEN",
        "description": "Optional. Currency of contract value.",
        "allowed_values": "USD | PEN",
    },
    {
        "name": "start_date",
        "example": "2026-03-15",
        "description": "Optional. Project start date. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "expected_end_date",
        "example": "2026-09-15",
        "description": "Optional. Expected completion date. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "actual_end_date",
        "example": "",
        "description": "Optional. Actual completion date. Populated on completion. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "location",
        "example": "Lima",
        "description": "Optional. Region or city in Peru.",
        "allowed_values": "",
    },
    {
        "name": "notes",
        "example": "Subcontract under Consorcio Vial",
        "description": "Optional. Free text notes.",
        "allowed_values": "",
    },
]

QUOTES = [
    {
        "name": "project_code",
        "example": "PRY001",
        "description": "Required. Project code. Must exist in projects table.",
        "allowed_values": "Lookup → projects.project_code",
    },
    {
        "name": "entity_document_number",
        "example": "20612345678",
        "description": "Required. Document number of the quoting entity. Must exist in entities table.",
        "allowed_values": "Lookup → entities.document_number",
    },
    {
        "name": "date_received",
        "example": "2026-03-01",
        "description": "Required. Date the quote was received. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "title",
        "example": "Portland cement Type I x 42.5kg",
        "description": "Required. What was quoted.",
        "allowed_values": "",
    },
    {
        "name": "quantity",
        "example": "500",
        "description": "Optional. NUMERIC(15,4).",
        "allowed_values": "",
    },
    {
        "name": "unit_of_measure",
        "example": "bags",
        "description": "Optional. meters, units, hours, kg, bags, etc.",
        "allowed_values": "",
    },
    {
        "name": "unit_price",
        "example": "28.50",
        "description": "Optional. NUMERIC(15,4).",
        "allowed_values": "",
    },
    {
        "name": "subtotal",
        "example": "14250.00",
        "description": "Required. NUMERIC(15,2). quantity x unit_price or entered directly.",
        "allowed_values": "",
    },
    {
        "name": "igv_amount",
        "example": "2565.00",
        "description": "Optional. NUMERIC(15,2). IGV amount.",
        "allowed_values": "",
    },
    {
        "name": "total",
        "example": "16815.00",
        "description": "Required. NUMERIC(15,2). subtotal + igv_amount.",
        "allowed_values": "",
    },
    {
        "name": "currency",
        "example": "PEN",
        "description": "Required. Currency of the quote.",
        "allowed_values": "USD | PEN",
    },
    {
        "name": "exchange_rate",
        "example": "3.72",
        "description": "Required. PEN per USD rate at transaction date. NUMERIC(10,4).",
        "allowed_values": "",
    },
    {
        "name": "status",
        "example": "pending",
        "description": "Required. Quote status.",
        "allowed_values": "pending | accepted | rejected",
    },
    {
        "name": "document_ref",
        "example": "PRY001-QT-001",
        "description": "Optional. Document reference linking to SharePoint file.",
        "allowed_values": "",
    },
    {
        "name": "notes",
        "example": "Valid for 15 days",
        "description": "Optional. Reason for rejection or other context.",
        "allowed_values": "",
    },
]

COSTS = [
    # === HEADER SECTION — required fields first ===
    {
        "name": "document_ref",
        "example": "PRY001-AP-001",
        "description": "Required. Grouping key — rows with same value form one cost with multiple items.",
        "allowed_values": "",
    },
    {
        "name": "date",
        "example": "2026-03-15",
        "description": "Required. Date the expense occurred. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "title",
        "example": "Cement purchase for foundation work",
        "description": "Required. Overall invoice or expense title.",
        "allowed_values": "",
    },
    {
        "name": "bank_account",
        "example": "BCP-4567",
        "description": "Required. Bank account label. Must exist in bank_accounts table.",
        "allowed_values": "Lookup → bank_accounts.label",
    },
    {
        "name": "currency",
        "example": "PEN",
        "description": "Required. Currency of the cost.",
        "allowed_values": "USD | PEN",
    },
    {
        "name": "igv_rate",
        "example": "18",
        "description": "Required. IGV percentage. NUMERIC(5,2). Default 18.",
        "allowed_values": "Usually 18",
    },
    # === HEADER SECTION — optional fields ===
    {
        "name": "project_code",
        "example": "PRY001",
        "description": "Optional. Project code. Blank = SG&A. Derives cost_type automatically.",
        "allowed_values": "Lookup → projects.project_code",
    },
    {
        "name": "entity_document_number",
        "example": "20612345678",
        "description": "Optional. Supplier/vendor document number. Null if informal/unassigned.",
        "allowed_values": "Lookup → entities.document_number",
    },
    {
        "name": "exchange_rate",
        "example": "3.72",
        "description": "Optional. PEN per USD rate. Auto-looked up from exchange_rates table by date if blank.",
        "allowed_values": "",
    },
    {
        "name": "comprobante_type",
        "example": "factura",
        "description": "Optional. Type of payment document.",
        "allowed_values": "factura | boleta | recibo_por_honorarios | liquidacion_de_compra | planilla_jornales | none",
    },
    {
        "name": "comprobante_number",
        "example": "F001-00234",
        "description": "Optional. Payment document number.",
        "allowed_values": "",
    },
    {
        "name": "detraccion_rate",
        "example": "4",
        "description": "Optional. Detracción percentage. NUMERIC(5,2). Null if not applicable.",
        "allowed_values": "Varies by service type",
    },
    {
        "name": "payment_method",
        "example": "bank_transfer",
        "description": "Optional. How the payment was made.",
        "allowed_values": "bank_transfer | cash | check",
    },
    {
        "name": "quote_document_ref",
        "example": "PRY001-QT-001",
        "description": "Optional. Document ref of a prior accepted quote. Null if no prior quote.",
        "allowed_values": "Lookup → quotes.document_ref",
    },
    {
        "name": "due_date",
        "example": "2026-04-15",
        "description": "Optional. Payment due date. Feeds AP payment calendar. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "notes",
        "example": "Urgent delivery requested",
        "description": "Optional. Free-form context about the cost.",
        "allowed_values": "",
    },
    # === DETAIL SECTION — required fields first ===
    {
        "name": "item_title",
        "example": "Portland cement Type I x 42.5kg",
        "description": "Required. Line item title.",
        "allowed_values": "",
    },
    {
        "name": "category",
        "example": "materials",
        "description": "Required. Cost category. Different allowed values for project costs vs SG&A.",
        "allowed_values": "materials | labor | subcontractor | equipment_rental | permits_regulatory | software_licenses | partner_compensation | business_development | professional_services | office_admin | other",
    },
    {
        "name": "subtotal",
        "example": "14250.00",
        "description": "Required. NUMERIC(15,2). Line item total amount.",
        "allowed_values": "",
    },
    # === DETAIL SECTION — optional fields ===
    {
        "name": "quantity",
        "example": "500",
        "description": "Optional. NUMERIC(15,4). Null for lump sum lines.",
        "allowed_values": "",
    },
    {
        "name": "unit_of_measure",
        "example": "bags",
        "description": "Optional. meters, units, hours, kg, days, bags, etc.",
        "allowed_values": "",
    },
    {
        "name": "unit_price",
        "example": "28.50",
        "description": "Optional. NUMERIC(15,4). Null for lump sum lines.",
        "allowed_values": "",
    },
]

AR_INVOICES = [
    {
        "name": "project_code",
        "example": "PRY001",
        "description": "Required. Project code. Must exist in projects table.",
        "allowed_values": "Lookup → projects.project_code",
    },
    {
        "name": "bank_account",
        "example": "Interbank-7890",
        "description": "Required. Bank account label. Must exist in bank_accounts table.",
        "allowed_values": "Lookup → bank_accounts.label",
    },
    {
        "name": "entity_document_number",
        "example": "20198765432",
        "description": "Required. Document number of the client entity.",
        "allowed_values": "Lookup → entities.document_number",
    },
    {
        "name": "partner_company_name",
        "example": "Korakuen Ingeniería S.A.C.",
        "description": "Required. Name of the partner company that issued the invoice.",
        "allowed_values": "Lookup → partner_companies.name",
    },
    {
        "name": "invoice_number",
        "example": "F001-00089",
        "description": "Required. Own invoice numbering from Alegra/Contasis.",
        "allowed_values": "",
    },
    {
        "name": "comprobante_type",
        "example": "factura",
        "description": "Required. Type of payment document. Always factura for construction AR.",
        "allowed_values": "factura | boleta | recibo_por_honorarios",
    },
    {
        "name": "invoice_date",
        "example": "2026-04-01",
        "description": "Required. Invoice issue date. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "due_date",
        "example": "2026-05-01",
        "description": "Optional. Payment due date. Format: YYYY-MM-DD.",
        "allowed_values": "",
    },
    {
        "name": "subtotal",
        "example": "150000.00",
        "description": "Required. NUMERIC(15,2). Invoice subtotal amount.",
        "allowed_values": "",
    },
    {
        "name": "igv_rate",
        "example": "18",
        "description": "Required. IGV percentage. NUMERIC(5,2). Default 18.",
        "allowed_values": "Usually 18",
    },
    {
        "name": "detraccion_rate",
        "example": "4",
        "description": "Optional. Detracción percentage. NUMERIC(5,2). Null if not applicable.",
        "allowed_values": "Varies by service type",
    },
    {
        "name": "retencion_applicable",
        "example": "false",
        "description": "Required. Whether client will withhold retención. BOOLEAN.",
        "allowed_values": "true | false",
    },
    {
        "name": "retencion_rate",
        "example": "3",
        "description": "Optional. Retención percentage. NUMERIC(5,2). Required if retencion_applicable is true. Default 3.",
        "allowed_values": "Usually 3 (if applicable)",
    },
    {
        "name": "currency",
        "example": "PEN",
        "description": "Required. Currency of the invoice.",
        "allowed_values": "USD | PEN",
    },
    {
        "name": "exchange_rate",
        "example": "3.72",
        "description": "Required. PEN per USD rate at transaction date. NUMERIC(10,4).",
        "allowed_values": "",
    },
    {
        "name": "document_ref",
        "example": "PRY001-AR-001",
        "description": "Optional. Document reference linking to SharePoint file.",
        "allowed_values": "",
    },
    {
        "name": "retencion_verified",
        "example": "false",
        "description": "Required. Manually set to true once confirmed that client paid retención to SUNAT. BOOLEAN. Default false.",
        "allowed_values": "true | false",
    },
    {
        "name": "notes",
        "example": "March 2026 billing",
        "description": "Optional. Free text notes.",
        "allowed_values": "",
    },
]


def main():
    print("Generating import templates...\n")

    templates = [
        ("entities.xlsx", ENTITIES),
        ("projects.xlsx", PROJECTS),
        ("quotes.xlsx", QUOTES),
        ("costs.xlsx", COSTS),
        ("ar_invoices.xlsx", AR_INVOICES),
    ]

    for filename, columns in templates:
        create_template(filename, columns)

    print(f"\nDone. {len(templates)} templates created in {TEMPLATES_DIR}/")


if __name__ == "__main__":
    main()

File: website/scripts/generate-templates.mjs
#!/usr/bin/env node
/**
 * Generate Excel import templates for the Korakuen system.
 * Run: node website/scripts/generate-templates.mjs
 *
 * Template structure (matches import-modal parser):
 *   Row 1: Column headers (snake_case keys)
 *   Row 2: Description
 *   Row 3: Example values
 *   Row 4: Valid values / constraints
 *   Row 5+: Data rows (user fills in)
 */

import XLSX from 'xlsx'
import { fileURLToPath } from 'url'
import { dirname, join } from 'path'

const __dirname = dirname(fileURLToPath(import.meta.url))
const OUT_DIR = join(__dirname, '..', 'public', 'templates')

function createTemplate(filename, columns, notes) {
  const headers = columns.map(c => c.key)
  const descriptions = columns.map(c => c.description)
  const examples = columns.map(c => c.example ?? '')
  const valids = columns.map(c => c.valid ?? '')

  const ws = XLSX.utils.aoa_to_sheet([headers, descriptions, examples, valids])

  // Set column widths based on longest content
  ws['!cols'] = columns.map(c => {
    const maxLen = Math.max(
      c.key.length,
      c.description.length,
      (c.example ?? '').toString().length,
      (c.valid ?? '').toString().length
    )
    return { wch: Math.min(Math.max(maxLen + 2, 12), 40) }
  })

  // Instructions sheet — structured table with all fields
  const instructionRows = [
    ['Data sheet structure:'],
    ['  Row 1: Column headers (do not modify)'],
    ['  Row 2: Column descriptions'],
    ['  Row 3: Example values'],
    ['  Row 4: Valid values / constraints'],
    ['  Row 5+: Enter your data here'],
    [],
    // Field reference table
    ['Field', 'Required', 'Description', 'Valid Values'],
    ...columns.map(c => [c.key, c.required ?? '', c.description, c.valid ?? '']),
    [],
    ...(notes ?? []).map(line => [line]),
  ]
  const wsInstructions = XLSX.utils.aoa_to_sheet(instructionRows)
  wsInstructions['!cols'] = [{ wch: 28 }, { wch: 44 }, { wch: 44 }, { wch: 60 }]

  const wb = XLSX.utils.book_new()
  XLSX.utils.book_append_sheet(wb, wsInstructions, 'Instructions')
  XLSX.utils.book_append_sheet(wb, ws, 'Data')
  XLSX.writeFile(wb, join(OUT_DIR, filename))
  console.log(`  ✓ ${filename}`)
}

// ============================================================
// Quotes Template (imports as pending invoices + invoice_items)
// ============================================================
createTemplate('quotes-template.xlsx', [
  { key: 'partner_name',            required: 'Yes',  description: 'Partner name',                   example: 'Korakuen SAC',     valid: 'Must exist in DB (tagged partner)' },
  { key: 'project_code',            required: 'Yes',  description: 'Project code',                  example: 'PRY001',           valid: 'Must exist in DB' },
  { key: 'entity_document_number',  required: 'Yes',  description: 'Supplier RUC/DNI',              example: '20123456789',      valid: 'Must exist in DB' },
  { key: 'date_received',           required: 'Yes',  description: 'Date received (quote date)',     example: '2026-03-15',       valid: 'YYYY-MM-DD' },
  { key: 'title',                   required: 'Yes',  description: 'Item title',                    example: 'Cement supply Q1' },
  { key: 'category',                required: 'No',   description: 'Item category',                 example: 'materials',        valid: 'Must exist in categories' },
  { key: 'quantity',                required: 'No',   description: 'Quantity',                       example: '100',              valid: 'Number' },
  { key: 'unit_of_measure',        required: 'No',   description: 'Unit of measure',                example: 'bags' },
  { key: 'unit_price',             required: 'No',   description: 'Unit price',                     example: '25.50',            valid: 'Number' },
  { key: 'subtotal',               required: 'Yes',  description: 'Subtotal (before IGV)',          example: '2550.00',          valid: 'Number, > 0' },
  { key: 'currency',               required: 'Yes',  description: 'Currency code',                  example: 'PEN',              valid: 'USD, PEN' },
  { key: 'exchange_rate',          required: 'Yes',  description: 'Exchange rate (PEN per USD)',     example: '3.72',             valid: '2.5–6.0' },
  { key: 'status',                 required: 'Yes',  description: 'Quote status',                   example: 'pending',          valid: 'pending, accepted, rejected' },
  { key: 'document_ref',           required: 'No',   description: 'Document reference',             example: 'COT-001' },
  { key: 'notes',                  required: 'No',   description: 'Notes',                          example: '' },
], [
  'Notes:',
  '  - Each row creates a quote (pending invoice) + one invoice_item',
  '  - Quotes with status "pending" or "rejected" are invisible to financial views',
  '  - Only "accepted" quotes become real invoices visible in totals, balances, and settlement',
  '  - You can change quote status from the Prices page after import',
  '  - partner_name must match an entity tagged as "partner" in the database',
  '  - project_code and entity_document_number must already exist in the database',
  '  - If quantity and unit_price are provided: subtotal must equal quantity × unit_price',
  '  - igv_amount and total are derived automatically (18% IGV)',
])

// ============================================================
// Invoices Template
// ============================================================
createTemplate('invoices-template.xlsx', [
  { key: 'direction',              required: 'Yes',                          description: 'Invoice direction',                      example: 'payable',          valid: 'payable, receivable' },
  { key: 'partner_name',           required: 'Yes',                          description: 'Partner name',                           example: 'Korakuen SAC',     valid: 'Must exist in DB' },
  { key: 'invoice_date',           required: 'Yes',                          description: 'Invoice date',                           example: '2026-03-15',       valid: 'YYYY-MM-DD' },
  { key: 'title',                  required: 'No',                           description: 'Invoice title (header)',                 example: 'Road materials' },
  { key: 'invoice_number',         required: 'No',                           description: 'Invoice number',                         example: 'F001-00123' },
  { key: 'currency',               required: 'Yes',                          description: 'Currency code',                          example: 'PEN',              valid: 'USD, PEN' },
  { key: 'igv_rate',               required: 'Yes',                          description: 'IGV rate (%)',                             example: '18',               valid: '0 or 18' },
  { key: 'exchange_rate',          required: 'Yes',                          description: 'Exchange rate (PEN per USD)',             example: '3.72',             valid: '2.5–6.0' },
  { key: 'project_code',           required: 'Yes, for receivable',          description: 'Project code',                           example: 'PRY001',           valid: 'Must exist in DB' },
  { key: 'entity_document_number', required: 'Yes, for receivable',          description: 'Entity RUC/DNI',                         example: '20123456789',      valid: 'Must exist in DB' },
  { key: 'comprobante_type',       required: 'Yes',                          description: 'Comprobante type',                       example: 'factura',          valid: 'factura, boleta, recibo_por_honorarios, liquidacion_de_compra, planilla_jornales, none, pending' },
  { key: 'document_ref',           required: 'Yes, for payable',             description: 'Document reference (grouping key)',      example: 'PRY001-AP-001' },
  { key: 'due_date',               required: 'No',                           description: 'Due date',                               example: '2026-04-15',       valid: 'YYYY-MM-DD' },
  { key: 'detraccion_rate',        required: 'No',                           description: 'Detracción rate (%)',                    example: '12',               valid: '0–100, PEN invoices only' },
  { key: 'retencion_applicable',   required: 'No',                           description: 'Retención applies?',                     example: 'false',            valid: 'true/false, receivable only' },
  { key: 'retencion_rate',         required: 'No',                           description: 'Retención rate (%). Defaults to 8% if retencion_applicable=true', example: '8', valid: '0–100, receivable only' },
  { key: 'notes',                  required: 'No',                           description: 'Notes',                                  example: '' },
  { key: 'item_title',             required: 'Yes',                          description: 'Line item title',                        example: 'Cement bags' },
  { key: 'category',               required: 'Yes, for payable',             description: 'Line item category',                     example: 'materials',        valid: 'materials, labor, subcontractor, equipment_rental, housing_food, other, software_licenses, partner_compensation, professional_services, other_sga' },
  { key: 'subtotal',               required: 'Yes',                          description: 'Line item subtotal (before IGV)',        example: '5000.00',          valid: 'Number, > 0' },
  { key: 'quantity',               required: 'No',                           description: 'Line item quantity',                     example: '200',              valid: 'Number' },
  { key: 'unit_of_measure',        required: 'No',                           description: 'Unit of measure',                        example: 'bags' },
  { key: 'unit_price',             required: 'No',                           description: 'Unit price',                             example: '25.00',            valid: 'Number' },
], [
  'Notes:',
  '  - Rows with the same document_ref are grouped into one invoice with multiple line items',
  '  - partner_name must match an entity tagged as "partner" in the database',
  '  - Detracción only applies to PEN invoices',
  '  - Retención only applies to receivable invoices',
  '  - If retencion_applicable=true and no retencion_rate is provided, defaults to 8%',
])

// ============================================================
// Payments Template (handles both invoice payments and direct transactions)
// ============================================================
createTemplate('payments-template.xlsx', [
  { key: 'invoice_document_ref',   required: 'No',                                            description: 'Links payment to an existing invoice. Leave blank to create a direct transaction',  example: 'PRY001-AP-001', valid: 'Must exist in DB if provided' },
  { key: 'direction',              required: 'Yes',                                            description: 'Cash flow direction',                          example: 'outbound',          valid: 'inbound, outbound' },
  { key: 'partner_name',           required: 'Yes',                                            description: 'Partner name',                                 example: 'Korakuen SAC',      valid: 'Must exist in DB' },
  { key: 'title',                  required: 'No',                                             description: 'What appears on bank app. Defaults: Pago/Cobro/Detraccion/Retencion', example: 'Pago cemento', valid: 'Free text' },
  { key: 'payment_date',           required: 'Yes',                                            description: 'Payment date',                                 example: '2026-03-20',        valid: 'YYYY-MM-DD' },
  { key: 'amount',                 required: 'Yes',                                            description: 'Payment amount',                               example: '5000.00',           valid: 'Number, > 0' },
  { key: 'currency',               required: 'Yes',                                            description: 'Currency code',                                example: 'PEN',               valid: 'USD, PEN' },
  { key: 'exchange_rate',          required: 'No',                                             description: 'Exchange rate (PEN per USD). Auto-filled from DB if blank', example: '3.72', valid: '2.5–6.0' },
  { key: 'payment_type',           required: 'No',                                             description: 'Payment type. Defaults to regular if blank',   example: 'regular',           valid: 'regular, detraccion, retencion' },
  { key: 'operation_number',        required: 'Yes, except retencion',                          description: 'Bank operation/transaction number (numero de operacion)',           example: '00123456',          valid: 'Free text' },
  { key: 'bank_account',           required: 'Yes, for invoice regular/detraccion payments',   description: 'Bank account in BankName-Last4 format. Currency must match payment', example: 'BCP-1234', valid: 'Must exist in DB' },
  { key: 'entity_document_number',  required: 'No (direct transactions only)',                  description: 'Supplier/client RUC/DNI for direct transactions', example: '20123456789',    valid: 'Must exist in DB if provided' },
  { key: 'project_code',           required: 'No (direct transactions only)',                  description: 'Project code for direct transactions',         example: 'PRY001',            valid: 'Must exist in DB' },
  { key: 'category',               required: 'Yes, for outbound direct transactions',          description: 'Cost category for direct transactions',        example: 'materials',         valid: 'Must match cost_type (project_cost if project_code given, sga otherwise)' },
  { key: 'document_ref',           required: 'No',                                             description: 'Payment receipt reference',                    example: 'PRY001-PY-001' },
  { key: 'notes',                  required: 'No',                                             description: 'Notes',                                       example: '' },
], [
  'Notes:',
  '  - If invoice_document_ref is provided: links payment to an existing invoice',
  '  - If invoice_document_ref is blank: creates a direct transaction (auto-generates invoice + payment)',
  '  - partner_name must match an entity tagged as "partner" in the database',
  '  - Retención only applies to receivable invoices',
  '  - exchange_rate is auto-filled from the database if left blank',
])

console.log(`\nAll templates written to: ${OUT_DIR}`)
